### ISSUE TITLE
Large number of emissive triangles cause Vulkan errors

### ISSUE THREAD

**Issue by KnightTec** [Lachei/VulkanPBRT#18]
> What error is occurring? The size per light is 8 * 4 * sizeof(float) = 128Byte/Light -> ~120KB for 1000 lights We should easily be able to add 10000 lights -> ~1MByte.
![Screenshot (398)](https://user-images.githubusercontent.com/38377295/144281678-f1b71689-d2bc-4eab-b752-f0f9e65828f6.png)
![Screenshot (399)](https://user-images.githubusercontent.com/38377295/144281693-da371b44-6267-4c9e-9f02-194865ae957d.png)

Those errors occured on my system when roughly more than 1000 lights were added. At 1500 lights the errors only appeared atfter a few frames and at 7000 lights no frame at all would be rendered.

_Originally posted by @KnightTec in https://github.com/Lachei/VulkanPBRT/issues/16#issuecomment-983858582_

**Comment by Lachei** [comment]
Looks to me as if the rendering time gets to high and thus the rendering job is cancelled without begin finished and thus the queue submission might not be marked as done...
Will test a large number of emissive triangles tomorrow morning...

**Comment by Lachei** [comment]
Just checked with  over 2000 triangle lights on a rtx 2070. Besides low framerates no problem seem to have occured.
I would suggest to add a define which can switch the ligth sampling from importance sampling over the light strength on the surface(current solution) and a uniform sampling for a large amount of lights. 
This would then avoid a complete search of the light array.
Aditionally to reduce variance we should consider a sampling of the light only based on the light strength. For this we have to store overall light strength of all lights together and an inclusive summation for the lights to then be able to search binary for the chosen light. Specifically the light sampling is then as follows:
```
sampleLight:
    float chosenLightStrength = randomFloat() * lightStrengthSum
    int lightIndex = binarySearch(lights, chosenLightStrength)
    return lightIndex
```
However for the beginning uniform sampling is sufficient i think.

**Comment by Lachei** [comment]
I will have time to implement this tomorrow.

**Comment by Lachei** [comment]
"Fix" for this should be implemented with #20.
At 800 area light sources the implementation automatically switches to a uniform sampling of the area lights. This reduces ray tracing times dramatically, however degrades convergence speed.
In the future the setting which sampling method should be used might get forwarded to a command line argument.
